### PR TITLE
Install julia modules

### PR DESCRIPTION
**Pull request by ocefpaf**
Needs #131

This is not ready for review yet. I need to adapt the CI ahnd ensure this will work with bot conda and pip installs.

### CODE REVIEW

**Comment by ocefpaf** [comment]
@aaronspring, sorry, I did not see you merged the main branch here and ended up force pushing. Usually I force push in a draft PR. I can change that and squash in the end if that is preferable.

**Comment by aaronspring** [comment]
Do what works best for you. I don’t mind

**Comment by ocefpaf** [comment]
Looks like it is working. A Julia environment is created at `xbitinfo-0.0.3.dev11+g531de33.d20220901/Project.toml` (right version!) and the first call to `install()` creates the environment and install everything. The subsequent calls do not install but they are a bit slow b/c julia checks for everything again. I'll see if I can make that a it faster.

```ipython
In [3]: from xbitinfo.julia_helpers import install

In [4]: install()
[ Info: Julia version info
Julia Version 1.6.5
Commit 9058264a69* (2021-12-19 12:30 UTC)
Platform Info:
  OS: Linux (x86_64-conda-linux-gnu)
  uname: Linux 5.15.0-47-generic #51-Ubuntu SMP Thu Aug 11 07:51:15 UTC 2022 x86_64 x86_64
  CPU: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz: 
                 speed         user         nice          sys         idle          irq
       #1-12  2600 MHz      33910 s        126 s      10668 s     766226 s          0 s
       
  Memory: 15.441070556640625 GB (1106.19921875 MB free)
  Uptime: 6815.34 sec
  Load Avg:  0.72  0.78  0.63
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, skylake)
Environment:
  JULIA_SSL_CA_ROOTS_PATH = /opt/conda/envs/TEST/ssl/cacert.pem
  JULIA_DEPOT_PATH = /opt/conda/envs/TEST/share/julia:
  JULIA_PROJECT_BACKUP = 
  JULIA_DEPOT_PATH_BACKUP = 
  JULIA_LOAD_PATH_BACKUP = 
  JULIA_SSL_CA_ROOTS_PATH_BACKUP = 
  JULIA_SSL_CA_ROOTS_PATH = /opt/conda/envs/TEST/ssl/cacert.pem
  TERM = xterm
  JULIA_DEPOT_PATH = /opt/conda/envs/TEST/share/julia:
  PATH = /opt/conda/envs/TEST/bin:/opt/conda/condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/home/conda/bin
  HOME = /home/conda
  JULIA_DEPOT_PATH_BACKUP = 
  JULIA_LOAD_PATH_BACKUP = 
  JULIA_SSL_CA_ROOTS_PATH_BACKUP = 
[ Info: Julia executable: /opt/conda/envs/TEST/bin/julia
[ Info: Trying to import PyCall...
┌ Info: PyCall is already installed and compatible with Python executable.
│ 
│ PyCall:
│     python: /opt/conda/envs/TEST/bin/python3.10
│     libpython: /opt/conda/envs/TEST/lib/libpython3.10.so.1.0
│ Python:
│     python: /opt/conda/envs/TEST/bin/python3.10
└     libpython: 
  Activating environment at `/opt/conda/envs/TEST/share/julia/environments/xbitinfo-0.0.3.dev11+g531de33.d20220901/Project.toml`
   Resolving package versions...
  No Changes to `/opt/conda/envs/TEST/share/julia/environments/xbitinfo-0.0.3.dev11+g531de33.d20220901/Project.toml`
  No Changes to `/opt/conda/envs/TEST/share/julia/environments/xbitinfo-0.0.3.dev11+g531de33.d20220901/Manifest.toml`
   Resolving package versions...
  No Changes to `/opt/conda/envs/TEST/share/julia/environments/xbitinfo-0.0.3.dev11+g531de33.d20220901/Project.toml`
  No Changes to `/opt/conda/envs/TEST/share/julia/environments/xbitinfo-0.0.3.dev11+g531de33.d20220901/Manifest.toml`
/opt/conda/envs/TEST/lib/python3.10/site-packages/xbitinfo/julia_helpers.py:43: UserWarning: It is recommended to restart Python so that the Julia environment is properly initialized.
  warnings.warn(
Out[4]: True
```

PS: no idea on what is going on with Windows and RTD. I'll try to address them later.

**Comment by aaronspring** [comment]
RTD is sometimes buggy- don’t bother here

**Comment by ocefpaf** [comment]
Windows is failing with:

```
julia.core.JuliaError: Exception 'could not load library "libgit2"
```

that is probably b/c I'm stalling the Julia dependencies from a tagged git release in the helper function. I'm not sure how to fix that b/c it is probably a limitation of the Julia installed in the GHA Windows VM. I could try to not pull them from the GitHub released tag but confess I prefer that option b/c it allows us for better version control.

I'll see if I can fix this in the CI...

**Comment by ocefpaf** [comment]
> RTD is sometimes buggy- don’t bother here

The issue there was that I forgot to remove the external Julia install call. Windows seems to also pass now that I'm using a GHA to provide Julia instead of the default one.

PS: I'd love for someone who uses only pip packages to test this a bit more. I'm particularly concerned with a possible slowdown in the import times.

**Comment by observingClouds** [comment]
@ocefpaf, I just tried the pip installation with all dependencies coming from pypi and I could only see a slight difference in timing. The import took about 45s with the currently on Conda-forge available Conda-package and about 50s with this PR installed via pip. Why are you concerned that the import times might be longer? Would you mind explaining the reasoning behind that?

**Comment by ocefpaf** [comment]
> @ocefpaf, I just tried the pip installation with all dependencies coming from pypi and I could only see a slight difference in timing. The import took about 45s with the currently on Conda-forge available Conda-package and about 50s with this PR installed via pip. Why are you concerned that the import times might be longer? Would you mind explaining the reasoning behind that?

The first run would should be similar to an "install time" but the second one should be faster. However, it is not as fast as something that is already installed b/c pyjulia will try to install the Julia modules, check if they exists first and, b/c they exists, run its course. So, in theory, that check if the julia modules are installed, is new and should add some overhead. Probably not much that we notice. To be honest I'm quite new to Julia and I not even 100% sure that this is exactly what is happening under the hood.

**Comment by observingClouds** [comment]
Thanks for the explanation! Is there anything keeping this from being merged? I would turn make the install silent for the users, but otherwise it looks good to me. Although it is kind of annoying that the package will have longer import times with the new approach, the possibility to install it via pip and conda are a huge benefit.

File: xbitinfo/julia_helpers.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        rev="v0.6.0",
+    )
+    Main.eval(f"Pkg.add(bitinformation_spec, {io_arg})")
+
+    Main.statsbase_spec = Main.PackageSpec(
+        name="StatsBase",
+        url="https://github.com/JuliaStats/StatsBase.jl",
+        rev="v0.33.21",
+    )
+    Main.eval(f"Pkg.add(statsbase_spec, {io_arg})")
```
Comment (mkitti) [inline xbitinfo/julia_helpers.py:125]:
```suggestion
    Main.bitinformation_spec = Main.PackageSpec(
        name="BitInformation",
        url="https://github.com/milankl/BitInformation.jl",
        rev="v0.6.0",
    )
    Main.statsbase_spec = Main.PackageSpec(
        name="StatsBase",
        url="https://github.com/JuliaStats/StatsBase.jl",
        rev="v0.33.21",
    )
    Main.eval(f"Pkg.add([bitinformation_spec, statsbase_spec], {io_arg})")
```

Comment (mkitti) [reply xbitinfo/julia_helpers.py:125]: We can consolidate the two `Pkg.add` calls into a single call.

Comment (ocefpaf) [reply xbitinfo/julia_helpers.py]: @mkitti thanks for looking into this PR! Please let us know if you have more suggestions.

**Comment by ocefpaf** [comment]
> Thanks for the explanation! Is there anything keeping this from being merged?

I don't think so. Let's see if the CIs are happy with the latest commits.

> I would turn make the install silent for the users

Done!

> but otherwise it looks good to me. Although it is kind of annoying that the package will have longer import times with the new approach, the possibility to install it via pip and conda are a huge benefit.

I'll make some tests and come up with some numbers to see how bad that is. Also, there is an active development on pysr that I indent to port here as soon as it stabilizes. It should make things even easier.

**Review by observingClouds — APPROVED** [review]

**Comment by observingClouds** [comment]
> I'll make some tests and come up with some numbers to see how bad that is. Also, there is an active development on pysr that I indent to port here as soon as it stabilizes. It should make things even easier.

I'll merge this for now as it makes the package installable via conda and pip. Just open a new PR for further refinements. Thanks @ocefpaf for all the work!!

**Comment by aaronspring** [comment]
This is great. Thanks for helping us out @ocefpaf @mkitti

**Comment by ocefpaf** [comment]
> This is great. Thanks for helping us out @ocefpaf @mkitti

Thank you for this awesome package! In the long run it is probably worth investing in a C, C++, or (even better ( a Rust core that we can wrap in Python, Julia, and R. That would make the code more robust and tested by multiple communities, and it would be much easier to package ;-p

**Pull request merged**